Content.MapRenderer/Painters/GridPainter.cs

```diff
@@ -75,29 +92,63 @@ public void Run(Image gridCanvas, IMapGrid grid)
                         $"No sprite component found on an entity for which a server sprite component exists. Prototype id: {prototype.ID}");
                 }
 
-                var xOffset = 0;
-                var yOffset = 0;
-                var tileSize = 1;
-
                 var transform = _sEntityManager.GetComponent<TransformComponent>(entity);
                 if (_cMapManager.TryGetGrid(transform.GridID, out var grid))
                 {
-                    xOffset = (int) Math.Abs(grid.LocalBounds.Left);
-                    yOffset = (int) Math.Abs(grid.LocalBounds.Bottom);
-                    tileSize = grid.TileSize;
-                }
+                    var position = transform.LocalPosition;
 
-                var position = transform.LocalPosition;
-                var x = ((float) Math.Floor(position.X) + xOffset) * tileSize * TilePainter.TileImageSize;
-                var y = ((float) Math.Floor(position.Y) + yOffset) * tileSize * TilePainter.TileImageSize;
-                var data = new EntityData(sprite, x, y);
+                    var (x, y) = TransformLocalPosition(position, grid);
+                    var data = new EntityData(sprite, x, y);
 
-                components.GetOrAdd(transform.GridID, _ => new List<EntityData>()).Add(data);
+                    components.GetOrAdd(transform.GridID, _ => new List<EntityData>()).Add(data);
+                }
             }
 
             Console.WriteLine($"Found {components.Values.Sum(l => l.Count)} entities on {components.Count} grids in {(int) stopwatch.Elapsed.TotalMilliseconds} ms");
 
             return components;
         }
+
+        private ConcurrentDictionary<GridId, List<DecalData>> GetDecals()
+        {
+            var stopwatch = new Stopwatch();
+            stopwatch.Start();
+
+            var decals = new ConcurrentDictionary<GridId, List<DecalData>>();
```

Comment:
This doesn't need to be a concurrent dictionary without any parallel loops writing into it. Far as I know it should also be safe to iterate it across threads later if it's not modified during that time.

Reply:
It is yeah (PVS makes extensive use of it); concurrentdictionary is also slow as balls.